src/compiler/checker.ts

```diff
@@ -27313,7 +27317,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 return convertAutoToAny(flowType);
             }
         }
-        else if (!assumeInitialized && !containsUndefinedType(type) && containsUndefinedType(flowType)) {
+        else if (!assumeInitialized && !containsUndefinedType(type) && !containsVoidType(type) && containsUndefinedType(flowType)) {

```

Comment:
The problem is that the `typeof` expression is the antecedent of the `value` variable here and it changes the `void` type somewhat unexpectedly.

[narrowTypeByLiteralExpression](https://github.dev/microsoft/TypeScript/blob/44152bc22edcf05d45faa06eae8aea2d81433ae4/src/compiler/checker.ts#L26530-L26534) discards `void` from `string | void` type for the `assumeTrue === false` case here.

So when we get to this line here we have:
```ts
type // string | void
flowType // string | undefined
``` 

As the other antecedent pushes `undefined` into the `antecedentTypes` in the [getTypeAtFlowBranchLabel](https://github.dev/microsoft/TypeScript/blob/44152bc22edcf05d45faa06eae8aea2d81433ae4/src/compiler/checker.ts#L26062).

I don't think this can be changed since this is how TS behaves today:
```ts
declare const a: string | void

if (typeof a === 'undefined') {
  a // undefined
}

if (typeof a !== 'undefined') {
  a // string
}
```

So checking against `voidType` here was my best shot at fixing this.